src/util/address.rs

```diff
@@ -511,13 +512,59 @@ impl Address {
         }
     }
 
-    /// Create a pay to taproot address
-    pub fn p2tr(taptweaked_key: schnorrsig::PublicKey, network: Network) -> Address {
+
+    /// Create a pay to taproot address for a given internal key and merkle root of tascript,
+    /// performing required tweaks according to BIP-342.
+    ///
+    /// Taproot (BIP-342) requires the keys to be tweaked with either tapscript merkle root or
+    /// self-tweaked with its own tagged hash. This function constructs non-tapscript based P2TR
+    /// address from an original untweaked public key value ("internal key"), performing the
+    /// necessary tweak with the merkle root of a tapscript.
+    ///
+    /// For generating P2TR addresses based on single public key please use
+    /// [`Address::p2tr_internal_key`] and [`Address::p2tr_tweaked_key`] methods.
+    #[inline]
+    pub fn p2tr_script<C: Verification>(secp: &Secp256k1<C>, mut internal_key: schnorrsig::PublicKey, merkle_root: TapBranchHash, network: Network) -> Address {
+        internal_key.script_tweak(&secp, merkle_root);
+        Address::p2tr_tweaked_key(internal_key, network)
+    }
+
+    /// Create a pay to taproot address for a given single key, tweaking the key value with its own
+    /// tagged hash according to BIP-342.
+    ///
+    /// Taproot (BIP-342) requires the keys to be tweaked with either tapscript merkle root or
+    /// self-tweaked with its own tagged hash. This function constructs non-tapscript based P2TR
+    /// address from an original untweaked public key value ("internal key"), performing the
+    /// necessary self-tweak.
+    ///
+    /// For generating P2TR addresses based on an already tweaked public key please use
+    /// [`Address::p2tr_tweaked_key`] method.
+    ///
+    /// For generating P2TR addresses with tapscript please use [`Address::p2tr_script`] method.
+    #[inline]
+    pub fn p2tr_internal_key<C: Verification + Signing>(secp: &Secp256k1<C>, mut internal_key: schnorrsig::PublicKey, network: Network) -> Address {
```

Comment:
I think there should only be one API. `new_p2tr(secp, internal_key, Option<merkle_root>, network)`